mypy/semanal.py

```diff
@@ -2432,6 +2451,32 @@ def can_be_type_alias(self, rv: Expression, allow_none: bool = False) -> bool:
                 return True
         return False
 
+    @contextmanager
+    def basic_type_applications_set(self, s: AssignmentStmt) -> Iterator[None]:
+        old = self.basic_type_applications
+        self.basic_type_applications = self.can_possibly_be_indexed_alias(s)
+        try:
+            yield
+        finally:
+            self.basic_type_applications = old
+
+    def can_possibly_be_indexed_alias(self, s: AssignmentStmt) -> bool:
+        """Like can_be_type_alias(), but simpler and doesn't require analyzed rvalue.
+
+        Instead, use lvalues/annotations structure to figure out whether this can
+        potentially be a type alias definition.
+        """
+        if len(s.lvalues) > 1:
+            return False
+        if not isinstance(s.lvalues[0], NameExpr):
+            return False
+        if s.unanalyzed_type is not None and not self.is_pep_613(s):
```

Comment:
Shouldn't we always return True if is_pep_613(s) returns True?

It may also be reasonable to require explicit `TypeAlias` for recursive aliases.

Reply:
> Shouldn't we always return True if is_pep_613(s) returns True?

We are interested only in aliases that can be recursive, so as a micro-optimization we only do this for r.h.s that can be recursive. I will add a comment on this.
> It may also be reasonable to require explicit `TypeAlias` for recursive aliases.

Hm, yeah this will simplify logic a bit. OTOH it may be hard to reliably detect cases where it was intended and missing, to give a good error message. I will keep the current logic as is, it is easy to change later, as this part is decoupled from everything else.